website/blog/2024-12-04-from-natural-language-to-sql/index.md

```diff
@@ -0,0 +1,1067 @@
+    sql_code: str = Field(description="The SQL code block")
+
+def get_sql_gen_chain():
+    """Set up the SQL generation chain."""
+    sql_gen_prompt = ChatPromptTemplate.from_messages(
+        [
+            (
+                "system",
+                """You are an SQL assistant with expertise in SQL query generation. \n
+Answer the user's question based on the provided documentation snippets and the database schema provided below. Ensure any SQL query you provide is valid and executable. \n
+Structure your answer with a description of the query, followed by the SQL code block. Here are the documentation snippets:\n{retrieved_docs}\n\nDatabase Schema:\n
+- Customers(CustomerID, CustomerName, ContactName, Address, City, PostalCode, Country)
+- Orders(OrderID, CustomerID, OrderDate)
+- OrderDetails(OrderDetailID, OrderID, ProductID, Quantity)
+- Products(ProductID, ProductName, Price)\n
+Here is the user's question:""",
+            ),
+            ("placeholder", "{messages}"),
+        ]
+    )
+
+    # Initialize the OpenAI LLM
+    llm = ChatOpenAI(temperature=0, model="gpt-4")
+
+    # Create the code generation chain
+    sql_gen_chain = sql_gen_prompt | llm.with_structured_output(SQLQuery)
+
+    return sql_gen_chain
+```
+
+### **Node Descriptions**
+#### **1. Translate Input**
+
+The `translate_input` node translates user queries into English to standardize processing, ensuring all inputs are compatible with downstream nodes.
+
+**Purpose:** Normalize user input to English for consistent processing.
+
+**Process:**
+
+- Extracts the latest user input from the state.
+
+- Uses the `ChatOpenAI` model to translate the input or verify it’s already in English.
+
+- Updates the state with the translated input.
+
+**Key Considerations:**
+
+- Maintains original intent and context during translation.
+
+- Provides accurate results even for multilingual queries.
+
+**Examples:**
+
+- Input: _"Quantos pedidos foram realizados em Novembro?"_
+
+- Translated: _"How many orders were made in November?"_
+
+- Input: _"Combien de ventes avons-nous enregistrées en France ?"_
+
+- Translated: _"How many sales did we record in France?"_
+
+**Code:**
+
+```python
+def translate_input(state: GraphState):
+    print("---TRANSLATING INPUT---")
+    messages = state["messages"]
+    user_input = messages[-1][1]  # Get the latest user input
+
+    # Translation prompt for the model
+    translation_prompt = f"""
+Translate the following text to English. If the text is already in English, repeat it exactly without any additional explanation.
+
+Text:
+{user_input}
+"""
+    # Call the OpenAI LLM to translate the text
+    translated_response = llm.invoke(translation_prompt)
+    translated_text = translated_response.content.strip()  # Access the 'content' attribute and strip any extra spaces
+    state["translated_input"] = translated_text  # Save the translated input
+    print(f"Translated Input: {translated_text}")
+
+    return state
+```
```

Comment:
Does translating to english before translating to SQL improve performance? Or would it work just as well to translate from whatever the input language is to SQL? Worth motivating this step.